### PR TITLE
Give six readers one vocabulary instead of six guesses (KBR-174)

### PR DESCRIPTION
**Pull request by shelpuk**
## Context for the Product Owner

Kitty Bridge sits between a coding agent and an AI provider and translates between vendor formats. Its test harness exists to prove nothing is **lost or altered** in that translation — that is the product's core promise.

The harness proves this by having one "reader" per vendor format convert each request into a single common shape, then comparing. **That only works if the readers describe the same thing with the same words.** When two readers disagree on a name, the harness reports a difference that isn't real — and a false alarm in the tool that guards correctness is worse than no tool, because it trains people to ignore it.

Two readers exist today and **they had already drifted**: one called an attached file `document`, the other called it `file`. Four more readers are still to be written, each of which would have guessed again.

This change moves three such rules out of comments and into enforced code, so the remaining four readers cannot invent new answers. No product behaviour changes — this is entirely test infrastructure — but it protects the reliability of the checks that gate every future release.

Closes KBR-174.

---

## What was pinned

| Rule | Was | Now |
|---|---|---|
| Decoding a tool call's arguments | One private copy in one reader | `contract.decode_arguments()` |
| Naming content the grammar can't model | Prose; two readers already disagreed | `contract.OPAQUE_ALIASES` + `opaque_kind()`, enforced by `Opaque` |
| Fingerprinting that content | Prose; two different recipes in use | `contract.opaque_digest()` and `contract.text_digest()` |

**An attached file is now `document`.** Anthropic *and* Bedrock Converse both spell it that way on the wire — confirmed against AWS's `bedrock-runtime` 2023-09-30 service model, not from memory — so exactly one reader had to move rather than two.

**The fingerprint has two recipes, deliberately.** A block is hashed from its canonical JSON; content identified by text is hashed from its text. One recipe was considered and dropped on a checked fact: Chat Completions carries a refusal as a **bare string** with no block to hash, so a single rule would force that reader to invent a wrapper — reintroducing this exact defect one level down.

**A camelCase vendor type raises rather than being auto-converted.** Shipping a name-conversion routine with no caller and no test data would be a second source of drift, not a cure for one. The reader that first meets real ones (Bedrock, nine of them) writes it then, with a corpus in hand. A *new snake_case* vendor type still passes through and projects normally, so a vendor release is not a harness outage.

## A correction to the design document

`TEST_SUITE.md` §7.4.1 told authors to defer this shared vocabulary "to the first reader that needs one". That advice was wrong, and the `document`/`file` split is what it cost. The section now records why: **a shared rule deferred to the first consumer is really deferred to the second**, because the first has already answered it alone and moved on.

## Verification

- Full gate green — **4858 passed**, 2 pre-existing skips, 0 failures.
- `ruff check`, `mypy src/kitty`, `lint-imports` (5 contracts) all clean.
- **17-mutation falsification sweep**, each injected alone and reverted.

Sixteen mutations went red immediately. **The seventeenth survived, and was right to.** The test for one guard used an input that a *different* guard rejected first, so the guard it named had no test at all — a green suite, a correct-sounding test name, and dead code underneath. Fixed and re-verified. That is the same class of defect this ticket exists to prevent, caught on this ticket's own work.

## Found and deliberately left out

Three more rules that are declared but unchecked, filed rather than folded in: KBR-191, KBR-192, KBR-193. KBR-193 is written as a judgement call rather than a defect, since already-planned work may cover it. KBR-179 also stays separate by agreement.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01J2sAEXym49SsSKWadFLDqf

### CODE REVIEW

**Comment by shelpuk** [comment]
## Code review applied — `b3e38a0`

The review returned **REQUEST CHANGES** on one finding and raised eight more. All nine are addressed. Summary for anyone reading this later, because the headline finding is the interesting part.

### The critical one: an acceptance criterion with no test that could fail

`test_a_kind_that_is_not_a_string_is_a_TypeError` was scenery. I deleted the guard it names and the suite stayed green at 810 passed — because every parameter raises an *incidental* `TypeError` further down: `fullmatch` on an int, `fullmatch` with a bytes pattern, `.get()` on an unhashable list. The code was correct. The test proved nothing.

**This was the third instance of one defect inside this change.** My own falsification sweep caught the first (the snake_case guard was shadowed by the alias branch, which raises the same type). The review caught the second and third. The rule is now written into the test and the commit:

> When two guards raise the same exception type, `pytest.raises(T)` cannot tell them apart. Pin the message, or pick an input the first guard cannot claim.

Fixed with `match="must be a str"`, and verified: deleting the guard now goes red.

### The one with a real failure mode behind it

`opaque_kind` had the same hole plus a worse problem. `OPAQUE_ALIASES.get(["a"])` raised `TypeError: unhashable type` **from the lookup itself, before any guard ran**. A reader's `except ValueError` cannot catch that, so it would escape as a raw crash instead of `UnreadableBodyError` — the exact membership defect this harness has shipped once before. It also raised `ValueError` for a non-`str` where `Opaque` raises `TypeError`, so D12's split held in one place only.

Guard now runs **before** the lookup, and raises `TypeError`.

### Two comments that were themselves the defect this PR exists to delete

- `OPAQUE_ALIASES` claimed its per-format list was "held honest by `test_contract.py`". **Nothing read it.** An unchecked claim about checking, inside a change whose whole thesis is that unchecked claims are comments. Now says plainly that only the Responses row is derived from the reader's own sets and the rest is a starting point, not an authority.
- §7.4.1 described `kind` as "converted to snake_case" two lines above the paragraph explaining that it deliberately does **not** convert. Reworded.

### Falsification: 22 rounds now

Five added, all red:

| | mutation | why it exists |
|---|---|---|
| 18 | `Opaque`'s type guard deleted | the critical finding |
| 19 | `opaque_kind`'s type guard deleted | |
| 20 | that guard moved back **below** the lookup | proves the *ordering*, not just the guard |
| 21 | it raises `ValueError` instead of `TypeError` | proves D12's split |
| 22 | the `TypeError` message reworded | proves the `match=` is load-bearing, not decorative |

Rounds 20 and 22 exist only to stop this fix decaying back into scenery.

### Accepted as correct, not fixed

- **`ruff format --check`** reports 49 files on an unmodified `main` and has never been a gate. It was in my acceptance criteria in error; the criteria are corrected rather than the code. None of the 4 files I touched is among the 49.
- **`RecursionError`** on pathologically nested JSON — pre-existing in every body-level parse, no corpus produces it. The docstring's "never raises" is qualified rather than the behaviour changed.
- **A grep test** for the removed private helpers was named in my criteria and never written. The review confirmed both names are gone repo-wide; the clause is dropped rather than a low-value test added to satisfy it.

### Two intentional behaviour changes, confirmed

The review flagged these for confirmation and both are deliberate: an Anthropic block typed `file` now projects as `Opaque("document")` (D1), and a camelCase block type now aborts the projection with `UnreadableBodyError` where it previously projected as `Opaque` (D10). A new **snake_case** vendor type still projects normally — a vendor release is not a harness outage.

Full gate re-verified green; `ruff check`, `mypy src/kitty`, `lint-imports` all clean.

File: tests/harness/test_contract.py [hunk 1]
```diff
@@ -1158,6 +1158,306 @@ def read_request(self, captured: c.CapturedRequest) -> c.Request:
...
+        """
+        c.decode_arguments(raw, "p", {})
+
+    def test_the_decoded_mapping_is_not_the_caller_s_object(self) -> None:
+        """A reader must not be able to mutate a projection's arguments through the body it read."""
+        residual: dict[str, object] = {}
+
+        decoded = c.decode_arguments('{"a": 1}', "p", residual)
+
+        assert c.ToolUse(name="t", arguments=decoded).arguments == {"a": 1}
```
[the remark this answers is not in the document]
Comment (shelpuk) [reply tests/harness/test_contract.py]: **Resolved in `b3e38a0`** — this landed independently from the same finding in a local review pass, before I saw this comment.

Renamed to `test_the_decoded_mapping_is_accepted_by_ToolUse`, with a docstring that states plainly why the old name was wrong: `json.loads` always returns a fresh object, so the aliasing claim was true for a reason the test never exercised. Your read was exactly right.

I did not add the identity check you offered as the alternative. `decode_arguments` has no body dict to alias — it takes a `str` and returns `json.loads`'s output — so an identity assertion would have no second object to compare against and would be as decorative as the name it replaced. If the function ever grows a cache, that is the change that should bring the test.

Worth noting this was one of **three** instances of the same defect in this PR: a test naming one property while being satisfied by another. My falsification sweep caught the first; a reviewer caught the second; this is the third. The pattern is now written into the tests and the commit message — when two guards raise the same exception type, `pytest.raises(T)` cannot distinguish them, so pin the message or pick an input the first guard cannot claim.

File: tests/harness/reader_responses.py [hunk 2]
```diff
@@ -773,21 +773,24 @@ def _read_content_part(
...
 
         if kind == "input_image":
             return self._read_image(entry, path, residual)
 
+        # Named through the shared table, not restated: Anthropic and Converse
+        # both spell this `document`, and T-A3 first shipped `file` for it —
+        # one concept under two names is the delta no register row can claim.
         if kind == "input_file":
-            return c.Opaque("file")
+            return c.Opaque(c.opaque_kind("input_file"))
```
[the remark this answers is not in the document]
Comment (shelpuk) [reply tests/harness/reader_responses.py:793]: **Confirmed as a real gap, reproduced, and deliberately not fixed here. Thank you for pressing on "filed rather than overlooked" — that was the right question to ask.**

### It reproduces

```
file A  -> Opaque(kind='document', digest=None)     {"type":"input_file","file_id":"file-AAA","filename":"contract.pdf"}
file B  -> Opaque(kind='document', digest=None)     {"type":"input_file","file_id":"file-ZZZ","filename":"malware.pdf"}
identical? True                                      <-- a swapped file produces NO delta

refusal -> Opaque(kind='refusal', digest='9390298f...')
```

Two entirely different attached files project identically. And the last line is the sharper version of your point: **the same reader, one branch over, does carry a digest.** So the Responses reader is inconsistent with itself, not merely with the Anthropic reader.

### Scope is wider than this one line

`reader_responses.py` has four `Opaque` construction sites. Three carry no digest:

| site | what | digest |
|---|---|---|
| the 27 opaque item types (×2 call sites) | items | **none** |
| `refusal` | content part | yes |
| `input_file` (this line) | content part | **none** |

This is the KBR-179 defect class. I could not re-read KBR-179 to confirm it names the `input_file` **content part** specifically rather than only the **item** types — its summary says "item", and its Atlassian connector is down for my session. So I am not claiming it is already covered; I am flagging that all three sites need to be, and that will be verified on the ticket before it is worked.

### Why not fix it here

Out of scope by an explicit product-owner decision on this ticket (recorded as D6), and the reasoning survives contact with your finding:

1. **Proving the fix needs its own test** — that a swapped unmodelled item actually produces a delta. That is the assertion the defect is defined by, and it is more than the one-line change.
2. **A partial fix is worse than none.** Patching only `input_file` leaves the 27 item types blind and makes KBR-179 harder to reason about, not easier.
3. Adding a digest is a **behaviour change to a merged reader**; this PR is a vocabulary pin whose reader edits are renames and delegation.

### On "the gap is reinforced on both sides"

Fair, and I want to be precise about what I did. I changed that assertion from `Opaque("file")` to `Opaque("document")` — the canonical-name fix this ticket is about. I did not add the absent digest, so the test still pins `digest=None`. That is a pre-existing gap passing through an edited line, not one this PR introduces; but you are right that it now has my fingerprints on it, which is why it is being escalated rather than waved through.

### One correction to the suggested patch

`reader_responses.py` has no `_residualise` helper — that is the Anthropic reader's. Whoever takes KBR-179 will need the Responses reader's own bookkeeping, and must keep the top-level key in `consumed` when a nested value residualises, or `verify_total` reports `DroppedFieldsError` and names the wrong defect.
